File: 496.下一个更大元素-i.c
/*
 * @lc app=leetcode.cn id=496 lang=c
 *
 * [496] 下一个更大元素 I
 */

// @lc code=start
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *nextGreaterElement(int *nums1, int nums1Size, int *nums2, int nums2Size, int *returnSize)
{
  int i, top, *stack, *res, *hash;
  stack = (int *)malloc(sizeof(int) * nums2Size);
  res = (int *)malloc(sizeof(int) * nums1Size);
  hash = (int *)malloc(sizeof(int) * 10000);
  top = 0;
  stack[top++] = nums2[0];
  for (i = 0; i < nums1Size; i++)
    hash[nums1[i]] = i;

  for (i = 0; i < nums1Size; i++)
    res[i] = -1;

  for (i = 1; i < nums2Size; i++)
  {
    if (nums2[i] < stack[top - 1])
      stack[top++] = nums2[i];
    else if (nums2[i] == stack[top - 1])
      stack[top++] = nums2[i];
    else
    {
      while (top != 0 && nums2[i] > stack[top - 1])
      {
        if (hash[stack[top - 1]] >= 0)
          res[hash[stack[top - 1]]] = nums2[i];
        top--;
      }
      stack[top++] = nums2[i];
    }
  }
  *returnSize = nums1Size;
  return res;
}
// @lc code=end

File: store/496.下一个更大元素-i.js
/*
 * @lc app=leetcode.cn id=496 lang=javascript
 *
 * [496] 下一个更大元素 I
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function (nums1, nums2) {
  let map = new Map(),
    res = new Array(nums1.length).fill(-1),
    stack = [],
    top = 0;
  for (let [index, item] of nums1.entries()) map.set(item, index);

  stack[top++] = nums2[0];
  for (let i = 1, len = nums2.length; i < len; i++) {
    while (nums2[i] > stack[top - 1] && top !== 0) {
      if (map.has(stack[top - 1])) res[map.get(stack[top - 1])] = nums2[i];
      top--;
    }
    stack[top++] = nums2[i];
  }

  return res;
};
// @lc code=end
